### ISSUE TITLE
Episode length shorter then chunk length, can't pad the chunk

### ISSUE THREAD

**Issue by piojanu** [google-research/planet#14]
Hi!

I stumbled across another issue. I try to pad sequence in `chunk_sequence` function when it is too short to create chunk from it (sequence length < chunk length). I also return shorter chunk length for the last chunk if it is shorter (in this particular code it is always the only chunk, but I wrote it for general case anyway). I saw masking in overshooting calculation so I thought it should work, but there are some infinite values in divergence loss (if I understand it correctly). Here is the code and console out (it is debug config with intentionally set chunk length to too big value):

### Code
```python
def chunk_sequence(sequence, chunk_length, randomize=True, num_chunks=None):
  """Split a nested dict of sequence tensors into a batch of chunks.

  This function does not expect a batch of sequences, but a single sequence. A
  `length` key is added if it did not exist already. When `randomize` is set,
  up to `chunk_length - 1` initial frames will be discarded. Final frames that
  do not fit into a chunk are always discarded.

  Args:
    sequence: Nested dict of tensors with time dimension.
    chunk_length: Size of chunks the sequence will be split into.
    randomize: Start chunking from a random offset in the sequence,
        enforcing that at least one chunk is generated.
    num_chunks: Optionally specify the exact number of chunks to be extracted
        from the sequence. Requires input to be long enough.

  Returns:
    Nested dict of sequence tensors with chunk dimension.
  """

  with tf.device('/cpu:0'):
    if 'length' in sequence:
      length = sequence.pop('length')
    else:
      length = tf.shape(nested.flatten(sequence)[0])[0]
    needs_padding = tf.constant(False)

    if randomize:
      if num_chunks is None:
        num_chunks = tf.maximum(1, length // chunk_length - 1)
      else:
        num_chunks = num_chunks + 0 * length
      used_length = num_chunks * chunk_length

      needs_padding = length < used_length
      pad_len = used_length - length
      # TODO: Assert here that padding will be smaller then one chunk length
      sequence = tf.cond(needs_padding,
                         lambda: nested.map(
                             partial(_pad_tensor, length=pad_len, value=1), sequence),
                         lambda: sequence)
      max_offset = tf.cond(needs_padding,
                           lambda: 0 * length,
                           lambda: length - used_length)
      offset = tf.random_uniform((), 0, max_offset + 1, dtype=tf.int32)
    else:
      if num_chunks is None:
        num_chunks = length // chunk_length
      else:
        num_chunks = num_chunks + 0 * length
      used_length = num_chunks * chunk_length
      max_offset = 0
      offset = 0

    clipped = nested.map(
        lambda tensor: tensor[offset: offset + used_length],
        sequence)
    chunks = nested.map(
        lambda tensor: tf.reshape(
            tensor, [num_chunks, chunk_length] + tensor.shape[1:].as_list()),
        clipped)

    chunks['length'] = tf.cond(needs_padding,
                               lambda: tf.concat(
                                   [chunk_length * tf.ones((num_chunks - 1,), dtype=tf.int32),
                                    (chunk_length - pad_len) * tf.ones((1,), dtype=tf.int32)], 0),
                               lambda: chunk_length * tf.ones((num_chunks,), dtype=tf.int32))
    tf.print("Requested length: ", chunk_length, " True length", chunks['length'])
    tf.print("Rewards: ", chunks['reward'])
    tf.print("Actions: ", chunks['action'])
    
    return chunks
```

### Output

`[...]` means that out from above repeats in mostly unchanged form.

```
Rewards:  [[-0.1 -0.1 -0.1 ... 1 1 1]]
Requested length:  110  True length [100]
Actions:  [[[0.773366451 0.66181761 -0.936789095 ... 0.923150301 -0.953526735 0.0622075424]
  [-0.590437293 -0.892674208 0.174977049 ... -0.939423203 -0.186107203 -0.910980761]
  [-0.504322708 -0.61423862 -0.569634855 ... 0.924559832 -0.295185775 0.788345039]
  ...
  [1 1 1 ... 1 1 1]
  [1 1 1 ... 1 1 1]
  [1 1 1 ... 1 1 1]]]
[...]
Caused by op 'graph/CheckNumerics_5', defined at:
  File "/usr/local/anaconda3/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/anaconda3/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Users/piotr/Projects/Planning-in-Imagination/src/planet/planet/scripts/train.py", line 129, in <module>
    tf.app.run(lambda _: main(args_), remaining)
  File "/usr/local/anaconda3/lib/python3.6/site-packages/tensorflow/python/platform/app.py", line 125, in run
    _sys.exit(main(argv))
  File "/Users/piotr/Projects/Planning-in-Imagination/src/planet/planet/scripts/train.py", line 129, in <lambda>
    tf.app.run(lambda _: main(args_), remaining)
  File "/Users/piotr/Projects/Planning-in-Imagination/src/planet/planet/scripts/train.py", line 102, in main
    for unused_score in run:
  File "/Users/piotr/Projects/Planning-in-Imagination/src/planet/planet/training/running.py", line 199, in __iter__
    for value in self._process_fn(self._logdir, *args):
  File "/Users/piotr/Projects/Planning-in-Imagination/src/planet/planet/scripts/train.py", line 87, in process
    training.define_model, dataset, logdir, config):
  File "/Users/piotr/Projects/Planning-in-Imagination/src/planet/planet/training/utility.py", line 160, in train
    score, summary = model_fn(data, trainer, config)
  File "/Users/piotr/Projects/Planning-in-Imagination/src/planet/planet/training/define_model.py", line 95, in define_model
    os_posterior, os_mask, config.free_nats, debug=config.debug)
  File "/Users/piotr/Projects/Planning-in-Imagination/src/planet/planet/training/utility.py", line 211, in compute_losses
    losses[key] = tf.check_numerics(loss, key) if debug else loss
  File "/usr/local/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/gen_array_ops.py", line 817, in check_numerics
    "CheckNumerics", tensor=tensor, message=message, name=name)
  File "/usr/local/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/usr/local/anaconda3/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py", line 488, in new_func
    return func(*args, **kwargs)
  File "/usr/local/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3274, in create_op
    op_def=op_def)
  File "/usr/local/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1770, in __init__
    self._traceback = tf_stack.extract_stack()

InvalidArgumentError (see above for traceback): divergence : Tensor had Inf values
	 [[node graph/CheckNumerics_5 (defined at /Users/piotr/Projects/Planning-in-Imagination/src/planet/planet/training/utility.py:211)  = CheckNumerics[T=DT_FLOAT, _class=["loc:@graph/optimizer_main/gradients/graph/CheckNumerics_5_grad/CheckNumerics"], message="divergence", _device="/job:localhost/replica:0/task:0/device:CPU:0"](graph/Mean_5)]]
```

Do you have any hints that could possibly shorten my struggle? 😄 

Thanks for help!

**Comment by danijar** [comment]
Note that we're not using padding in PlaNet but instead randomly crop sub sequences of length `config.batch_shape[1]` from the collected episodes. If your collected sequences are at least of this length, this might work out of the box. I haven't tested training on variable length sequences though, so it's not a supported feature but if it doesn't work you might get it to work with minor changes.

**Comment by piojanu** [comment]
It is exactly what I'm trying to do. I'll try to root cause it on my own then.

**Comment by danijar** [comment]
Good luck, hope your project goes well!

**Comment by danijar** [comment]
@piojanu I looked into this a bit and parallel data collections seems to be working now. You can try this via `--params '{collect_batch_size: 5}'`.

**Comment by piojanu** [comment]
By parallel you mean batched collection in the collection step right? Not training and collecting in parallel?

**Comment by danijar** [comment]
Yes, exactly.

**Comment by piojanu** [comment]
Also, in the paper you've written that "we [...] add one additional episode to the data set every C update steps.". With batched collection you then add e.g. 5 new episodes in the collection step?

**Comment by danijar** [comment]
Yes, although this was not used in the paper.
